src/mainboard/starfive/visionfive2/bt0/src/dram.rs

```diff
@@ -0,0 +1,100 @@
+// NOTE: Those _could_ be an enum, but this makes it easily compatible with the
+// original C code.
+const F_SET: u8 = 1 << 0;
+const REG2G: u8 = 1 << 1;
+const REG4G: u8 = 1 << 2;
+const REG8G: u8 = 1 << 3;
+const REGSETALL: u8 = 1 << 4;
+const REGCLRSETALL: u8 = 1 << 5;
+const OFFSET_SEL: u8 = 1 << 6;
+
+struct MemCfg {
+    offset: u32,
+    mask: u32,
+    value: u32,
+    flags: u32,
+}
+
+impl MemCfg {
+    fn new(offset: u32, mask: u32, value: u32, flags: u32) -> Self {
+        MemCfg {
+            offset,
+            mask,
+            value,
+            flags,
+        }
+    }
+}
+
+/// Generate array of MemCfg from simply C-style struct array
+///
+/// Example:
+/// ```rs
+/// const MEM_CFG_0: [MemCfg; 3] = mem_cfg_arr![
+///     {0x0,   0x0, 0x00000001, REGSETALL},
+///     {0xf00, 0x0, 0x40001030, (OFFSET_SEL | F_SET | REG4G | REG8G)},
+///     {0xf00, 0x0, 0x40001030, (OFFSET_SEL | F_SET | REG2G)},
+/// ];
+/// ```
+///
+/// Results in:
+///
+/// ```rs
+/// const MEM_CFG_FOO: [MemCfg; 3] = [
+///     MemCfg {
+///         offset: 0x0,
+///         mask: 0x0,
+///         value: 0x00000001,
+///         flags: REGSETALL,
+///     },
+///     MemCfg {
+///         offset: 0xf00,
+///         mask: 0x0,
+///         value: 0x40001030,
+///         flags: (OFFSET_SEL | F_SET | REG4G | REG8G),
+///     },
+///     MemCfg {
+///         offset: 0xf00,
+///         mask: 0x0,
+///         value: 0x40001030,
+///         flags: (OFFSET_SEL | F_SET | REG2G),
+///     },
+/// ];
+/// ```
+macro_rules! mem_cfg_arr {
+    ($({ $o: expr, $m: expr, $v: expr, $f: expr }),* $(,)?) => {
+        [
+            $(MemCfg { offset: $o, mask: $m, value: $v, flags: $f }),*
```

Comment:
```suggestion
            $(MemCfg::new($o, $m, $v, $f as u32)),*
```

With the `MemCfg::new` being a `const fn` you could probably get rid of the macro. You would just have to cast the last argument to `u32` on each call below.

Reply:
Note that this was intermediate code.

At this point, we have working DRAM init.